File: C++/Baekjoon/Problem14567.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

int indegree[1001];
int res[1001];

int maxx(int a, int b) {
    if (a > b) return a;
    else return b;
}

int main() {
    vector<int> graph[1001];
    queue<int> q;

    int n, m;
    cin >> n >> m;
    for (int i = 0; i < m; i++) {
        int a, b;
        cin >> a >> b;
        graph[a].push_back(b);
        indegree[b]++;
    }

    for (int i = 1; i <= n; i++) {
        if (indegree[i] == 0) {
            q.push(i);
            res[i] = 1;
        }
    }

    while (!q.empty()) {
        int cur = q.front();
        q.pop();

        for (int i = 0; i < graph[cur].size(); i++) {
            int next = graph[cur][i];
            indegree[next]--;

            res[next] = maxx(res[next], res[cur] + 1);
            if (indegree[next] == 0) {
                q.push(next);
            }
        }
    }

    for (int i = 1; i <= n; i++) {
        cout << res[i] << " ";
    }
}

File: Python/Baekjoon/Problem14567.py
N, M = map(int, input().split())
graph = [[] for i in range(N + 1)]
indegree = [0 for i in range(N + 1)]
dp = [0 for i in range(N + 1)]
queue = []

for i in range(M):
    a, b = map(int, input().split())
    graph[a].append(b)
    indegree[b] += 1

for i in range(1, N + 1):
    if indegree[i] == 0:
        queue.append(i)
        dp[i] = 1

while queue:
    cur = queue[0]
    queue.pop(0)

    for i in range(len(graph[cur])):
        next = graph[cur][i]
        indegree[next] -= 1

        dp[next] = max(dp[next], dp[cur] + 1)
        if indegree[next] == 0:
            queue.append(next)

for i in range(1, N + 1):
    print(dp[i], end=' ')
